src/parse.rs

```diff
@@ -1195,16 +1203,35 @@ pub(crate) struct AlignmentIndex(usize);
 
 #[derive(Clone)]
 pub(crate) struct Allocations<'a> {
-    pub refdefs: HashMap<LinkLabel<'a>, LinkDef<'a>>,
+    pub refdefs: RefDefs<'a>,
     links: Vec<(LinkType, CowStr<'a>, CowStr<'a>)>,
     cows: Vec<CowStr<'a>>,
     alignments: Vec<Vec<Alignment>>,
 }
 
+/// Keeps track of the reference definitions defined in the document.
+#[derive(Clone)]
+pub struct RefDefs<'input>(pub(crate) HashMap<LinkLabel<'input>, LinkDef<'input>>);
+
+impl<'input, 'b, 's> RefDefs<'input>
+where
+    's: 'b,
```

Comment:
Is this bound necessary? I'm pretty sure it will make the output of `get()` unusable since it can't be used after the `self` borrow is released.

Reply:
It appears that it is necessary. And I think that it means something different: that the output of `get` cannot outlive `self` (not the borrow of self), which is indeed required for memory safety.

To confirm this, I added `let linkdef = parser.reference_definitions().get("a b c").unwrap();` to the test case and it compiles. That seems to confirm that the linkdef reference can indeed outlive the `RefDefs` reference.

Reply:
Try this:
```diff
diff --git a/src/parse.rs b/src/parse.rs
index 35e9647..3f8dfdf 100644
--- a/src/parse.rs
+++ b/src/parse.rs
@@ -1203,7 +1203,7 @@ pub(crate) struct AlignmentIndex(usize);
 
 #[derive(Clone)]
 pub(crate) struct Allocations<'a> {
-    pub refdefs: RefDefs<'a>,
+    pub refdefs: RefDefs<'a, 'a>,
     links: Vec<(LinkType, CowStr<'a>, CowStr<'a>)>,
     cows: Vec<CowStr<'a>>,
     alignments: Vec<Vec<Alignment>>,
@@ -1211,19 +1211,16 @@ pub(crate) struct Allocations<'a> {
 
 /// Keeps track of the reference definitions defined in the document.
 #[derive(Clone)]
-pub struct RefDefs<'input>(pub(crate) HashMap<LinkLabel<'input>, LinkDef<'input>>);
+pub struct RefDefs<'input, 'output>(pub(crate) HashMap<LinkLabel<'input>, LinkDef<'output>>);
 
-impl<'input, 'b, 's> RefDefs<'input>
-where
-    's: 'b,
-{
+impl<'input, 'output> RefDefs<'input, 'output> {
     /// Performs a lookup on reference label using unicode case folding.
-    pub fn get(&'s self, key: &'b str) -> Option<&'b LinkDef<'input>> {
+    pub fn get<'s>(&'s self, key: &'input str) -> Option<&'s LinkDef<'output>> {
         self.0.get(&UniCase::new(key.into()))
     }
 
     /// Provides an iterator over all the document's reference definitions.
-    pub fn iter(&'s self) -> impl Iterator<Item = (&'s str, &'s LinkDef<'input>)> {
+    pub fn iter<'s>(&'s self) -> impl Iterator<Item = (&'s str, &'s LinkDef<'output>)> {
         self.0.iter().map(|(k, v)| (k.as_ref(), v))
     }
 }
```
That
- Decouples the lifetime of `LinkLabel` from `LinkDef`
- Returns a `LinkDef` that can live longer than the `key`

I think previously this case would break:
```rust
let parser = unimplemented!();
let def = {
    let key = String::from("hello");
    parser.allocs.refdefs.get(&key)
};
```

Reply:
That said my example is sort of convoluted and it's probably ok to just not fix this.

Reply:
Ah, indeed that case breaks. But even with your patch, I cannot get it to compile. From my understanding, it still demands that they key lives as long as the hashmap keys.

It should obviously be possible to do this. I believe that `LinkLabel<'a>` should implement `Borrow<LinkLabel<'b>> where 'a: 'b` to enable it.

I have a lot more confidence in your lifetime analysis expertise, so please let me know when I'm missing something. I'll keep tinkering in the meantime to try to get your case to compile.

Reply:
Ok, I did more digging and asked for help on the xi zulip chat [here](https://xi.zulipchat.com/#narrow/stream/255911-rust/topic/hashmap.20get.20with.20temp.20keys/near/221519321). It seems that this is a limitation of the Hashmap API (and ultimately Rust itself).